include/antler/project/net_utils.hpp

```diff
@@ -0,0 +1,190 @@
+#pragma once
+
+/// @copyright See `LICENSE` in the root directory of this project.
+
+#include <antler/system/utils.hpp>
+
+#include <cstring>
+#include <iostream>
+#include <stdexcept>
+
+#include <curl/curl.h>
+#include <nlohmann/json.hpp>
+
+
+namespace antler::project {
+
+   // TODO replace later with string_view.ends_with() and string_view.starts_with()
+   static inline bool ends_with(std::string_view src, std::string_view comp) {
+      const std::size_t sz = comp.size();
+      if (src.size() < sz)
+         return false;
+      
+      return std::memcmp(&src[0] + src.size()-sz, &comp[0], sz) == 0;
+   }
+
+   static inline bool starts_with(std::string_view src, std::string_view comp) {
+      const std::size_t sz = comp.size();
+      if (src.size() < sz)
+         return false;
+      
+      return std::memcmp(&src[0], &comp[0], sz) == 0;
+   }
+
+   // object to encapsulate curl
+   struct curl {
+      inline curl() {
+         curl_global_init(CURL_GLOBAL_ALL);
+         curl_obj = curl_easy_init();
+         ANTLER_CHECK(curl_obj, "internal failure, curl_easy_init() failed");
+      }
+
+      ~curl() {
+         curl_easy_cleanup(curl_obj);
+      }
+
+      static std::size_t write_data(void* ptr, size_t sz, size_t nm, void* s) {
+         reinterpret_cast<std::string*>(s)->append(reinterpret_cast<char*>(ptr), sz*nm);
+         return sz*nm;
+      }
+
+      std::string authenticated_request(std::string_view token, std::string_view url) {
+         curl_easy_setopt(curl_obj, CURLOPT_XOAUTH2_BEARER, token.data());
+         return request(url);
+      }
+
+      std::string request(std::string_view url) {
+         curl_easy_setopt(curl_obj, CURLOPT_URL, url.data());
+         curl_easy_setopt(curl_obj, CURLOPT_WRITEFUNCTION, &curl::write_data);
+         curl_easy_setopt(curl_obj, CURLOPT_USERAGENT, "antler-proj");
+
+         std::string buff;
+
+         curl_easy_setopt(curl_obj, CURLOPT_WRITEDATA, &buff);
+
+         auto res = curl_easy_perform(curl_obj);
+         ANTLER_CHECK(!res, "internal failure, curl_easy_perform() failed with error {0}", curl_easy_strerror(res));
+         return buff;
+      }
+
+      inline static bool is_url(std::string_view url) { return starts_with(url, "http:") || starts_with(url, "https:"); }
+      inline bool is_reachable(std::string_view url) {
+         try {
+            (void)request(url);
+            return true;
+         } catch(...) {
+            return false;
+         }
+      }
+
+      CURL* curl_obj;
```

Comment:
Leaving this public feels like a latent defect waiting to happen.
```suggestion
private:
      CURL* curl_obj;
```